packages/jest-cli/src/generateEmptyCoverage.js

```diff
@@ -29,16 +32,15 @@ export default function(
     collectCoverageOnlyFrom: globalConfig.collectCoverageOnlyFrom,
   };
   if (Runtime.shouldInstrument(filename, coverageOptions, config)) {
-    // Transform file without instrumentation first, to make sure produced
-    // source code is ES6 (no flowtypes etc.) and can be instrumented
-    const transformResult = new Runtime.ScriptTransformer(
+    // Transform file with instrumentation to make sure initial coverage data is well mapped to original code.
+    const {code, mapCoverage, sourceMapPath} = new Runtime.ScriptTransformer(
       config,
-    ).transformSource(filename, source, false);
-    const instrumenter = createInstrumenter();
-    instrumenter.instrumentSync(transformResult.code, filename);
+    ).transformSource(filename, source, true);
+    const extracted = readInitialCoverage(code);
+
     return {
-      coverage: instrumenter.fileCoverage,
-      sourceMapPath: transformResult.sourceMapPath,
+      coverage: new FileCoverage(extracted.coverageData),
+      sourceMapPath: mapCoverage ? sourceMapPath : null,
```

Comment:
@jwbay I think `mapCoverage` indicates whether the coverageData should be mapped back to original code. If should, we return the `sourceMapPath`, and `coverage` will be transformed automatically. Otherwise, such as when we use babel with `babel-plugin-istanbul`, the coverageData is already for source map so we don't pass back the source map.